### PR TITLE
[IMP] website,* : reduce number of sql queries on generic code

### PR DESCRIPTION
**Pull request by rdeodoo**
Reduce number of queries on base method like dispatch, or base controller like /web/content 

Part of task-2211013

Following
    https://github.com/odoo/odoo/pull/46925
    https://github.com/odoo/odoo/pull/47257 

Most of these small queries are done in part of code called a lot of times.
Next step is to be more 'controller' specific, and reduce query / imp perf
for specific page /my /blog /forum /shop ...

Courtesy of Rco and AL for help and validation

Co-authored-by: Jérémy Kersten <jke@odoo.com>
Co-authored-by: Romain Derie <rde@odoo.com>

### CODE REVIEW

File: odoo/addons/base/models/ir_http.py [hunk 1]
```diff
@@ -318,7 +318,8 @@ def _get_record_and_check(self, xmlid=None, model=None, id=None, field='datas',
             # We have prefetched some fields of record, among which the field
             # 'write_date' used by '__last_update' below. In order to check
             # access on record, we have to invalidate its cache first.
-            record._cache.clear()
+            if not record.env.su:
+                record._cache.clear()
```
Comment (rdeodoo) [inline odoo/addons/base/models/ir_http.py]: This doesn't break https://github.com/odoo/odoo/commit/0d5fcdc1c7d13f3ac1d5868d7116b5c4e45a692c so might be a correct solution

Comment (rco-odoo) [reply odoo/addons/base/models/ir_http.py]: The `if` statement can even contain the statement below.

File: addons/website/tests/test_performance.py [hunk 2]
```diff
@@ -101,3 +101,9 @@ def test_40_perf_sql_queries_page_multi_level_menu(self):
         menu_aa.parent_id = menu_a
 
         self.assertEqual(self._get_url_hot_query(self.page.url), 14)
+
+    def test_50_perf_sql_web_content(self):
+        # assets route /web/content/..
+        assets_url = self.env['ir.attachment'].search([('url', '=like', '/web/content/%/web.assets_common.js')], limit=1).url
+        self._get_url_hot_query(assets_url)
+        self.assertEqual(self._get_url_hot_query(assets_url), 5)
```
Comment (rco-odoo) [inline addons/website/tests/test_performance.py]: Is there a reason for calling the method twice?

File: addons/website/models/ir_ui_view.py [hunk 3]
```diff
@@ -430,7 +430,7 @@ def _prepare_qcontext(self):
                 website=request.website,
                 url_for=url_for,
                 res_company=request.website.company_id.sudo(),
-                default_lang_code=request.env['ir.http']._get_default_lang().code,
+                # default_lang_code=request.env['ir.http']._get_default_lang().code,
```
Comment (rco-odoo) [inline addons/website/models/ir_ui_view.py]: remove it, no?

File: addons/website/models/website.py [hunk 4]
```diff
@@ -143,7 +143,8 @@ def _compute_menu(self):
                 # don't add child menu if parent is forbidden
                 if menu.parent_id and menu.parent_id in menus:
                     menu.parent_id._cache['child_id'] += (menu.id,)
-
+            # prefetch every website.page and ir.ui.view in once
```
Comment (rco-odoo) [inline addons/website/models/website.py]:
```suggestion
            # prefetch every website.page and ir.ui.view at once
```

File: odoo/addons/base/models/ir_http.py (hunk 1 above)
Comment (rco-odoo) [reply odoo/addons/base/models/ir_http.py]: Suggestion: you can actually explicitly call the check methods
```python
if not record.env.su:
    record.check_access_rights('read')
    record.check_access_rule('read')
```

File: odoo/addons/base/models/ir_http.py [hunk 5]
```diff
@@ -296,34 +296,39 @@ def _get_record_and_check(self, xmlid=None, model=None, id=None, field='datas',
...
+
+            # check read access
+            try:
+                # We have prefetched some fields of record, among which the field
+                # 'write_date' used by '__last_update' below. In order to check
+                # access on record, we have to invalidate its cache first.
+                if not record.env.su:
+                    record._cache.clear()
+                record.check_access_rights('read')
+                record.check_access_rule('read')
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_http.py]:
```suggestion
                if not record.env.su:
                    record.check_access_rights('read')
                    record.check_access_rule('read')
```

File: addons/website/tests/test_performance.py (hunk 2 above)
Comment (rdeodoo) [reply addons/website/tests/test_performance.py]: No it was leftover of dev as it was still in wip but good catch I removed it

File: odoo/addons/base/models/ir_http.py [hunk 6]
```diff
@@ -296,34 +296,38 @@ def _get_record_and_check(self, xmlid=None, model=None, id=None, field='datas',
...
+                    record = record_sudo
+
+            # check read access
+            try:
+                # We have prefetched some fields of record, among which the field
+                # 'write_date' used by '__last_update' below. In order to check
+                # access on record, we have to invalidate its cache first.
+                if not record.env.su:
+                    record._cache.clear()
+                record['__last_update']
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_http.py]: Why not this?
```suggestion
                if not record.env.su:
                    record.check_access_rights('read')
                    record.check_access_rule('read')
```

Comment (rdeodoo) [reply odoo/addons/base/models/ir_http.py]: I reverted that part for now to clean the PR and have a green runbot (tests are failing for reasons not related to the diff, such as wrong data page_id on menu etc).

But this change was not working and kept raising errors. I did not yet investigate further but I will come back to you once I do to move forward on this suggested change

Comment (rco-odoo) [reply odoo/addons/base/models/ir_http.py]: Thanks for the explanation!

File: addons/website/models/website.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 from odoo.addons.portal.controllers.portal import pager
 from odoo.http import request
 from odoo.modules.module import get_resource_path
-from odoo.osv.expression import FALSE_DOMAIN
+from odoo.osv.expression import AND, FALSE_DOMAIN
```
Comment (JKE-be) [inline addons/website/models/website.py]: to remove

File: addons/website/controllers/main.py [hunk 8]
```diff
@@ -489,5 +489,7 @@ def content_image(self, id=None, max_width=0, max_height=0, **kw):
 
     @http.route(['/favicon.ico'], type='http', auth='public', website=True, multilang=False, sitemap=False)
     def favicon(self, **kw):
-        # when opening a pdf in chrome, chrome tries to open the default favicon url
-        return self.content_image(model='website', id=str(request.website.id), field='favicon', **kw)
+        website = request.website
```
Comment (JKE-be) [inline addons/website/controllers/main.py]: add comment

**Comment by antonylesuisse** [comment]
r+

**Comment by JKE-be** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**